scraper/src/libretexts2zim/client.py

```diff
@@ -0,0 +1,135 @@
+import datetime
+import re
+from collections.abc import Callable
+
+import requests
+from bs4 import BeautifulSoup, NavigableString
+from pydantic import BaseModel
+
+from libretexts2zim.constants import logger
+
+HTTP_TIMEOUT_SECONDS = 15
+
+
+class LibreTextsParsingError(Exception):
+    pass
+
+
+class LibreTextsShelve(BaseModel):
+    title: str
+    content_url: str
+    image_url: str
+
+
+class LibreTextsHome(BaseModel):
+    welcome_text_paragraphs: list[str]
+    shelves: list[LibreTextsShelve]
+    welcome_image_url: str
+
+
+class LibreTextsMetadata(BaseModel):
+    """Metadata about a course."""
+
+    # Human readable name for the course.
+    name: str
+
+    # URL prefix for the course, e.g. for Geosciences which is at
+    # https://geo.libretexts.org/, the slug is `geo`
+    slug: str
+
+    def placeholders(
+        self, clock: Callable[[], datetime.date] = datetime.date.today
+    ) -> dict[str, str]:
+        """Gets placeholders for filenames.
+
+        Arguments:
+          clock: Override the default clock to use for producing the "period".
+        """
+
+        return {
+            "name": self.name,
+            "slug": self.slug,
+            "clean_slug": re.sub(r"[^.a-zA-Z0-9]", "-", self.slug),
+            "period": clock().strftime("%Y-%m"),
+        }
+
+
+class LibreTextsClient:
+    """Utility functions to read data from libretexts."""
+
+    def __init__(self, library_slug: str) -> None:
+        """Initializes LibreTextsClient.
+
+        Paremters:
+            library_url: Scheme, hostname, and port for the Libretext library
+                e.g. `https://geo.libretexts.org/`.
+        """
+        self.library_slug = library_slug
+
+    @property
+    def library_url(self) -> str:
+        return f"https://{self.library_slug}.libretexts.org/"
+
+    def _get_text(self, url: str) -> str:
+        """Perform a GET request and return the response as decoded text."""
+
+        logger.debug(f"Fetching {url}")
+
+        resp = requests.get(
+            url=url,
+            allow_redirects=True,
+            timeout=HTTP_TIMEOUT_SECONDS,
+        )
+        resp.raise_for_status()
+
+        return resp.text
+
+    def get_home(self) -> LibreTextsHome:
+        home_content = self._get_text(self.library_url)
+
+        soup = _get_soup(home_content)
+        return LibreTextsHome(
+            welcome_text_paragraphs=_get_welcome_text_from_home(soup),
+            welcome_image_url=_get_welcome_image_url_from_home(soup),
+            shelves=[],
+        )
+
+
+def _get_soup(content: str) -> BeautifulSoup:
+    return BeautifulSoup(content, "html.parser")
```

Comment:
What would be the advantage of using `lxml`? Isn't this parser a bit too "intelligent" for our use-case?

I think we need to hard-code this value to avoid surprises at runtime / ensure tests are testing the real thing.

Reply:
Speed, robustness, stability. I agree that we should not support multiple ones in a scraper scenario.